src/main/java/org/opensearch/neuralsearch/search/query/HybridCollectorManager.java

```diff
@@ -215,6 +219,139 @@ private DocValueFormat[] getSortValueFormats(final SortAndFormats sortAndFormats
+        // if collector doesn't have any hits we can just skip it and save some cycles by not doing merge
+        if (newTopDocs.totalHits.value == 0) {
+            return;
+        }
+        // we need to do actual merge because query result and current collector both have some score hits
+        TopDocsAndMaxScore originalTotalDocsAndHits = result.topDocs();
+        result.topDocs(mergeTopDocsAndMaxScores(originalTotalDocsAndHits, topDocsAndMaxScore), docValueFormats);
+    }
+
+    /**
+     * For collection of search results, return a single one that has results from all individual result objects.
+     * @param results collection of search results
+     * @return single search result that represents all results as one object
+     */
+    private ReduceableSearchResult reduceSearchResults(List<ReduceableSearchResult> results) {
+        return (result) -> {
+            for (ReduceableSearchResult r : results) {
+                // call reduce for results of each single collector, this will update top docs in query result
+                r.reduce(result);
+            }
+        };
+    }
+
+    @VisibleForTesting
+    protected TopDocsAndMaxScore mergeTopDocsAndMaxScores(TopDocsAndMaxScore source, TopDocsAndMaxScore newTopDocs) {
+        if (Objects.isNull(newTopDocs) || Objects.isNull(newTopDocs.topDocs) || newTopDocs.topDocs.totalHits.value == 0) {
+            return source;
+        }
+        // we need to merge hits per individual sub-query
+        // format of results in both new and source TopDocs is following
+        // doc_id | magic_number_1
+        // doc_id | magic_number_2
+        // ...
+        // doc_id | magic_number_2
+        // ...
+        // doc_id | magic_number_2
+        // ...
+        // doc_id | magic_number_1
+        ScoreDoc[] sourceScoreDocs = source.topDocs.scoreDocs;
+        ScoreDoc[] newScoreDocs = newTopDocs.topDocs.scoreDocs;
+
+        List<ScoreDoc> mergedScoreDocs = mergedScoreDocs(sourceScoreDocs, newScoreDocs, Comparator.comparing((scoreDoc) -> scoreDoc.score));
+        TotalHits mergedTotalHits = getMergedTotalHits(source, newTopDocs);
+        TopDocsAndMaxScore result = new TopDocsAndMaxScore(
+            new TopDocs(mergedTotalHits, mergedScoreDocs.toArray(new ScoreDoc[0])),
+            Math.max(source.maxScore, newTopDocs.maxScore)
+        );
+        return result;
+    }
+
+    /**
+     * Merge two score docs objects, result ScoreDocs[] object will have all hits per sub-query from both original objects.
+     * Logic is based on assumption that hits of every sub-query are sorted by score.
+     * Method returns new object and doesn't mutate original ScoreDocs arrays.
+     * @param sourceScoreDocs original score docs from query result
+     * @param newScoreDocs new score docs that we need to merge into existing scores
+     * @return merged array of ScoreDocs objects
+     */
+    private List<ScoreDoc> mergedScoreDocs(
```

Comment:
Can we move all this code to a new class for ex: HybridQueryMergeTopDocs ?

As I for sorting I am planning to create the similar class where I will keep all the merge related code?

```
/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
package org.opensearch.neuralsearch.search;

import java.util.Comparator;
import org.apache.lucene.search.FieldComparator;
import org.apache.lucene.search.FieldDoc;
import org.apache.lucene.search.Pruning;
import org.apache.lucene.search.ScoreDoc;
import org.apache.lucene.search.SortField;

public class HybridQueryMergeTopFieldDocs {
    final SortField[] sortFields;
    final FieldComparator<?>[] comparators;
    final int[] reverseMul;
    final Comparator<ScoreDoc> tieBreaker;

    public HybridQueryMergeTopFieldDocs(SortField[] sortFields, Comparator<ScoreDoc> tieBreaker) {
        this.sortFields = sortFields;
        this.tieBreaker = tieBreaker;
        comparators = new FieldComparator[sortFields.length];
        reverseMul = new int[sortFields.length];
        for (int compIDX = 0; compIDX < sortFields.length; compIDX++) {
            final SortField sortField = sortFields[compIDX];
            comparators[compIDX] = sortField.getComparator(1, Pruning.NONE);
            reverseMul[compIDX] = sortField.getReverse() ? -1 : 1;
        }
    }

    public boolean lessThan(final FieldDoc firstFD, final FieldDoc secondFD) {
        for (int compIDX = 0; compIDX < comparators.length; compIDX++) {
            final FieldComparator comp = comparators[compIDX];

            final int cmp = reverseMul[compIDX] * comp.compareValues(firstFD.fields[compIDX], secondFD.fields[compIDX]);

            if (cmp != 0) {
                return cmp < 0;
            }
        }
        return tieBreakLessThan(firstFD, secondFD, tieBreaker);
    }

    private boolean tieBreakLessThan(ScoreDoc firstDoc, ScoreDoc secondDoc, Comparator<ScoreDoc> tieBreaker) {
        assert tieBreaker != null;
        int value = tieBreaker.compare(firstDoc, secondDoc);
        return value < 0;
    }
}
```

Reply:
let's me think about proper LLD. Ideally we can do a strategy here and use implementation that is specific to the TopDocs type

Reply:
I pulled some code to a new Merger class. Not very match of abstracting is possible as ScoreDoc and SortField are not in the same class hierarchy, and case for general hybrid search is much simpler comparing to sorting. 
We can think and refactor more as a separate task some time later.